apps/s_client.c

```diff
@@ -2218,7 +2219,20 @@ int s_client_main(int argc, char **argv)
                            "Didn't find STARTTLS in server response,"
                            " trying anyway...\n");
             BIO_printf(sbio, "STARTTLS\r\n");
-            BIO_read(sbio, sbuf, BUFSIZZ);
+            mbuf_len = BIO_read(sbio, mbuf, BUFSIZZ);
+            if (mbuf_len < 0) {
+                BIO_printf(bio_err, "BIO_read failed\n");
+                goto end;
+            }
+            mbuf[mbuf_len] = '\0';
+            if (sscanf(mbuf, "%3d ", &numeric) != 1) {
+                BIO_printf(bio_err, "Server does not support STARTTLS.\n");
+                goto shut;
+            }
+            if (numeric != 382) {
```

Comment:
The number comes from [RFC 4642, section 2.2.1](https://tools.ietf.org/html/rfc4642#section-2.2.1) saying ("382 Continue with TLS negotiation"). Regarding "382 ", I was a) looking to `PROTO_IRC` b) the response code should be followed with a human readable text…but yes, it's should. Are you looking for something like that instead?

    mbuf_len = BIO_read(sbio, mbuf, BUFSIZZ);
    if (mbuf_len < 0) {
        BIO_printf(bio_err, "BIO_read failed\n");
        goto end;
    } else if (mbuf_len < 3) {
        BIO_printf(bio_err, "Server does not support STARTTLS.\n");
        goto shut;
    }
    mbuf[mbuf_len] = '\0';
    if (strncmp(mbuf, "382", 3) != 0) {
        BIO_printf(bio_err, "STARTTLS failed: %s", mbuf);
        goto shut;
    }